app/Http/Controllers/Maps/MapDataController.php

```diff
@@ -0,0 +1,799 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * @link       https://www.librenms.org
+ *
+ * @copyright  2019 Thomas Berberich
+ * @author     Thomas Berberich <sourcehhdoctor@gmail.com>
+ * @author     Steven Wilton <swilton@fluentit.com.au>
+ */
+
+namespace App\Http\Controllers\Maps;
+
+use App\Http\Controllers\Controller;
+use App\Models\AlertSchedule;
+use App\Models\Device;
+use App\Models\Link;
+use App\Models\Port;
+use App\Models\Service;
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Http\JsonResponse;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Log;
+use LibreNMS\Config;
+use LibreNMS\Util\Url;
+
+class MapDataController extends Controller
+{
+    protected static function geoLinks(Request $request)
+    {
+        $user = $request->user();
+
+        // Return a blank array for unknown link types
+        if ($request->link_type != 'xdp') {
+            return collect();
+        }
+
+        $linkQuery = Link::with('port', 'device', 'remoteDevice', 'device.location', 'remoteDevice.location')
+            ->whereHas('device', function (Builder $q) use ($user) {
+                $q->whereIn('status', [0, 1])
+                    ->where('disabled', 0)
+                    ->where('ignore', 0);
+
+                if (! $user->hasGlobalRead()) {
+                    $q->whereIntegerInRaw('device_id', \Permissions::devicesForUser($user));
+                }
+            })
+            ->whereHas('device.location', function (Builder $q) {
+                $q->whereNotNull('lat')
+                    ->whereNotNull('lng');
+            })
+            ->whereHas('remoteDevice', function (Builder $q) use ($user) {
+                $q->whereIn('status', [0, 1])
+                    ->where('disabled', 0)
+                    ->where('ignore', 0);
+
+                if (! $user->hasGlobalRead()) {
+                    $q->whereIntegerInRaw('device_id', \Permissions::devicesForUser($user));
+                }
+            })
+            ->whereHas('remoteDevice.location', function (Builder $q) {
+                $q->whereNotNull('lat')
+                    ->whereNotNull('lng');
+            })
+            ->whereHas('port', function (Builder $q) {
+                $q->where('ifOperStatus', 'up');
+            });
+
+        $group_id = $request->group;
+        if ($group_id) {
+            $linkQuery->join('device_group_device AS ldg', 'ld.device_id', '=', 'ldg.device_id')
```

Comment:
Is it possible to use relationships for this query?  That would avoid the phpstan errors.

Reply:
How does this change look?